### PR TITLE
Interfaces for add-ffi

### PR DESCRIPTION
**Pull request by jjaworska**
Interfaces for the "add-ffi" functionality

### CODE REVIEW

File: src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt
Comment (Giselus) [inline src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt:128]: I see that we have a problem with naming and we swapped meanings of declaration and definition. Foreign functions can only be declared, but not defined. We could probably fix it and swap all declarations and definitions, if you don't want to do it here you may keep it as it is for integrity (and maybe order someone else to fix naming 😈).

File: src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt
Comment (Giselus) [inline src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt:133]: We can't have own function declarations in code, but we can have foreignFunction declarations, so maybe we can add toString for printing. I'm unsure if foreign function declarations will be present in AST, so I don't know if my comment makes sense.

File: src/main/kotlin/cacophony/controlflow/functions/FunctionCallHandlerImpl.kt
Comment (Giselus) [inline src/main/kotlin/cacophony/controlflow/functions/FunctionCallHandlerImpl.kt:9]: If I understand, we'll have another implementation like _ForeignFunctionCallHandlerImpl_, so maybe _FunctionCallHandlerImpl_ can get a better name that shows that it's user created function? But I'm not sure, just a suggestion.

**Review by Giselus — APPROVED** [review]
Looks good overall. I approve, but please think if you want to change something according to comments.

File: src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt
Comment (jjaworska) [reply src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt:128]: that was actually quite bad. I changed the naming, but this makes the diff much bigger

File: src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt
Comment (jjaworska) [reply src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt:133]: Yes, definitely, I added both this and isEquivalent

File: src/main/kotlin/cacophony/controlflow/functions/FunctionCallHandlerImpl.kt
Comment (jjaworska) [reply src/main/kotlin/cacophony/controlflow/functions/FunctionCallHandlerImpl.kt:9]: I was thinking there was one FunctionCallHandlerImpl which just checks inside which function it uses. Maybe let's leave it up to decide for the person who will be implementing this?

File: src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt
Comment (jjaworska) [reply src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt:128]: but of course, it'll change snapshots once again :upside_down_face:

File: src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt
Comment (Giselus) [inline src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt:132]: Just a question, I see it's done this way in different places as well. Type.Functional already has returnType, so it seems like we have duplicate returnType here. Maybe one of them is not used, or something. It has too many occurrences in the code to check, so I'll just leave this comment and you may add this as an issue or maybe I don't understand something here.

**Review by Giselus — APPROVED** [review]
LGTM 👍

**Pull request merged**
